Sputnik upload

Pelevina opened the pull request.

Pelevina committed:
Add files via upload

Pelevina committed:
Add files via upload

Pelevina committed:
Update FileUploadController.java

Pelevina committed:
Delete src/main/java/com/plasmidEditor/sputnik/storage directory

Pelevina committed:
Add Upload Services

Pelevina committed:
Delete src/main/resources/templates directory

iporsh commented on src/main/java/com/plasmidEditor/sputnik/controllers/FileUploadController.java:
```diff
@@ -0,0 +1,46 @@
+		this.proteinFileUploadService = proteinFileUploadService;
+	}
+	
+	@PostMapping(path="/genbank/dna", produces=MediaType.APPLICATION_JSON_VALUE)
```
/genbank  можно использовать  для всей группы этотих контроллеров

Pelevina replied:
То есть получается можно использовать только ProteinFileUploadService?

iporsh replied:
нет))) 
`@Controller
@RequestMapping(value = "/genbank")
public class FileEditorController {`

Pelevina replied:
RequestMapping(value = "/genbank") теперь стоит перед классом контроллера

darina-ovis replied:
теперь путь будет выглядеть как /genbank/genbank/dna

iporsh commented on src/main/java/com/plasmidEditor/sputnik/controllers/FileUploadController.java:
```diff
@@ -0,0 +1,46 @@
+	@PostMapping(path="/genbank/dna", produces=MediaType.APPLICATION_JSON_VALUE)
+	public ResponseEntity<String> uploadDNAFile(@RequestParam("file") MultipartFile file) {
+		try {
+			dnaFileUploadService.upload(file);
```
MultipartFile это кастомное решение, не хорошо его пробрасывать дальше , лучше в твоем случае использовать InputStream

Pelevina replied:
Где оставить MultipartFile, а где заменить его на InputStream?

Pelevina replied:
Везде MultipartFile заменено на InputStream

iporsh replied:
на вход контролера должен быть MultipartFile
на вход сервисов InputStream

Pelevina replied:
Теперь на входе контроллера MultipartFile, а на входе сервисов InputStream

iporsh commented on src/main/java/com/plasmidEditor/sputnik/controllers/FileUploadController.java:
```diff
@@ -0,0 +1,46 @@
+			return ResponseEntity.ok("File was successfully uploaded");
+		}
+		catch(Exception e) {
+			return ResponseEntity.badRequest().body("File wasn't uploaded because " + e.getMessage());
```
А какой статус ответа будет приходить?

Pelevina replied:
В нормальном случае должен приходить 200, в случае ошибки - 400.

iporsh commented on src/main/java/com/plasmidEditor/sputnik/uploadServices/DNAFileUploadService.java:
```diff
@@ -0,0 +1,14 @@
+public class DNAFileUploadService implements FileUploadService<DNASequence> {
+
+	@Override
+	public void upload(MultipartFile file) {
```
MultipartFile это кастомное решение, не хорошо его пробрасывать дальше , лучше в твоем случае использовать InputStream

Pelevina replied:
Везде MultipartFile заменено на InputStream

iporsh commented on src/main/java/com/plasmidEditor/sputnik/uploadServices/FileUploadService.java:
```diff
@@ -0,0 +1,7 @@
+import org.springframework.web.multipart.MultipartFile;
+
+public interface FileUploadService<T> {
+	public void upload(MultipartFile file);
```
MultipartFile это кастомное решение, не хорошо его пробрасывать дальше , лучше в твоем случае использовать InputStream

Pelevina replied:
Везде MultipartFile заменено на InputStream

lyzhib committed:
Update DNAFileUploadService.java

lyzhib committed:
Update FileUploadService.java

lyzhib committed:
Update ProteinFileUploadService.java

darina-ovis commented on src/main/java/com/plasmidEditor/sputnik/controllers/FileUploadController.java:
```diff
@@ -0,0 +1,46 @@
+			return ResponseEntity.ok("File was successfully uploaded");
+		}
+		catch(Exception e) {
+			return ResponseEntity.badRequest().body("File wasn't uploaded because " + e.getMessage());
```
объявлено, что метод возвращает json, но возвращаемая строка - это не json объект. Часто для методов на создание объектов используется статус 201 created и при этом пустое тело ответа. В случае ошибки может прийти статус 400 и json объект. Например, {"error":"failed to upload....", "reason": "some reason"} В вашем случае ошибка тоже вернется строкой. Посмотрите в сторону аннотации ResponseBody https://www.baeldung.com/spring-request-response-body

Pelevina replied:
Сделан внутренний класс, который должен преобразовать ответ в json.

darina-ovis commented on src/main/java/com/plasmidEditor/sputnik/uploadServices/DNAFileUploadService.java:
```diff
@@ -0,0 +1,14 @@
+
+	@Override
+	public void upload(MultipartFile file) {
+		// TODO Auto-generated method stub
```
уберите плз мусорный код

Pelevina replied:
Убрано лишнее

darina-ovis commented on src/main/java/com/plasmidEditor/sputnik/uploadServices/FileUploadService.java:
```diff
@@ -0,0 +1,47 @@
+
+public interface FileUploadService<T> {
+	public void upload(MultipartFile file);
+	@Autowired
+	public FileUploadController(StorageService storageService) {
+		this.storageService = storageService;
+	}
+
```
с этим интерфейсом явно что-то не то. Это у вас сервис или контроллер?

Pelevina replied:
Сейчас этого нет

darina-ovis commented on src/main/java/com/plasmidEditor/sputnik/uploadServices/FileUploadService.java:
```diff
@@ -0,0 +1,47 @@
+		this.storageService = storageService;
+	}
+
+	@GetMapping("/")
+	public String listUploadedFiles(Model model) throws IOException {
+
+		model.addAttribute("files", storageService.loadAll().map(
+				path -> MvcUriComponentsBuilder.fromMethodName(FileUploadController.class,
+						"serveFile", path.getFileName().toString()).build().toUri().toString())
+				.collect(Collectors.toList()));
+
+		return "uploadForm";
+	}
+
+	@GetMapping("/files/{filename:.+}")
+	@ResponseBody
+	public ResponseEntity<Resource> serveFile(@PathVariable String filename) {
+
+		Resource file = storageService.loadAsResource(filename);
+		return ResponseEntity.ok().header(HttpHeaders.CONTENT_DISPOSITION,
+				"attachment; filename=\"" + file.getFilename() + "\"").body(file);
+	}
+
```
Назначение этих методов мне не понятно. Даже если допустить, что вы хотели написать контроллер для получения файлов, то зачем вам модель? К тому же, экспорт файлов - это отдельная задача и должна быть в отдельном пулл реквесте

Pelevina replied:
Сейчас этого нет

darina-ovis commented on src/main/java/com/plasmidEditor/sputnik/uploadServices/FileUploadService.java:
```diff
@@ -0,0 +1,47 @@
+		return "redirect:/";
+	}
+
+	@ExceptionHandler(StorageFileNotFoundException.class)
+	public ResponseEntity<?> handleStorageFileNotFound(StorageFileNotFoundException exc) {
+		return ResponseEntity.notFound().build();
+	}
```
хорошее решение для обработки ошибок, только этому методу место в контроллере, т.е. **классу** с аннотацией **RestController** (или Controller)

Pelevina replied:
Сейчас этого нет в этом файле

Pelevina replied:
ExceptionHandler теперь в контроллере

darina-ovis commented on src/main/java/com/plasmidEditor/sputnik/uploadServices/ProteinFileUploadService.java:
```diff
@@ -0,0 +1,24 @@
+import org.biojava.nbio.core.sequence.ProteinSequence;
+import org.springframework.web.multipart.MultipartFile;
+
+public class ProteinFileUploadService implements FileUploadService<ProteinSequence> {
+
+	@Override
+	public void upload(MultipartFile file) {
+		// TODO Auto-generated method stub
+		void init();
+
+	void store(MultipartFile file);
+
+	Stream<Path> loadAll();
+
+	Path load(String filename);
+
+	Resource loadAsResource(String filename);
+
+	void deleteAll();
+	}
+
+}
```
Вот это больше похоже на интерфейс, чем interface FileUploadService<T>

Pelevina replied:
Сейчас этого нет

darina-ovis commented on src/main/java/com/plasmidEditor/sputnik/uploadServices/DNAFileUploadService.java:
```diff
@@ -0,0 +1,14 @@
+	@Override
+	public void upload(MultipartFile file) {
+		// TODO Auto-generated method stub
+		SpringApplication.run(UploadingFilesApplication.class, args);
```
Не понимаю, что вы пытаетесь тут сделать. У нас должно быть одно Spring приложение. Нам нужен класс с аннотацией RestController с точками входа (методы аннотированные PostMapping, GetMapping и т.д.) и из этих методов вызываются методы сервиса, где сервис выполняет непостредственно логику. Например, загрузку файла, проверку файла, сохранение в базу данных и прочее.

Pelevina replied:
Сейчас этого нет

darina-ovis commented on src/main/java/com/plasmidEditor/sputnik/uploadServices/FileUploadService.java:
```diff
@@ -0,0 +1,47 @@
+public interface FileUploadService<T> {
+	public void upload(MultipartFile file);
+	@Autowired
+	public FileUploadController(StorageService storageService) {
+		this.storageService = storageService;
+	}
```
Что за storageService?

Pelevina replied:
Сейчас этого нет

Pelevina committed:
Update FileUploadService.java

Pelevina committed:
Update DNAFileUploadService.java

Pelevina committed:
Update ProteinFileUploadService.java

Pelevina committed:
Add ExceptionHandler

Pelevina committed:
Замена MultipartFile на InputStream

Pelevina committed:
Замена MultipartFile на InputStream

Pelevina committed:
Замена MultipartFile на InputStream

Pelevina committed:
Update DNAFileUploadService.java

Pelevina committed:
Замена MultipartFile на InputStream

Pelevina committed:
Update FileUploadController.java

darina-ovis commented on src/main/java/com/plasmidEditor/sputnik/controllers/FileUploadController.java:
```diff
@@ -0,0 +1,51 @@
+	}
+	
+	@PostMapping(path="/genbank/dna", produces=MediaType.APPLICATION_JSON_VALUE)
+	public ResponseEntity<JsonError> uploadDNAFile(@RequestParam("file") InputStream file) {
+		dnaFileUploadService.upload(file);
+		return new ResponseEntity<>(HttpStatus.CREATED);
+	}
+	
+	@PostMapping(path="/genbank/protein", produces=MediaType.APPLICATION_JSON_VALUE)
+	public ResponseEntity<JsonError> uploadProteinFile(@RequestParam("file") InputStream file) {
+		proteinFileUploadService.upload(file);
+		return new ResponseEntity<>(HttpStatus.CREATED);
+	}
```
Вам не нужно возвращать ошибку при успешном выполнении метода. Не должно быть тут ResponseEntity<JsonError>
Для статуса можно использовать аннтотацию ResponseStatus

Pelevina replied:
Использована ResponseStatus перед контроллером. Лишний /genbank убран

darina-ovis commented on src/main/java/com/plasmidEditor/sputnik/uploadServices/DNAFileUploadService.java:
```diff
@@ -0,0 +1,14 @@
+
+	@Override
+	public void upload(InputStream file) {
+		
```
Даже если бд не готова, вы можете сделать интерфейс для dao с пустыми методами либо с сохранением файла на диск (возможная реализация). Уровень же сервиса можно написать

Pelevina replied:
Есть логика

darina-ovis replied:
Я бы хотела видеть тут логику, одинаковую и для сохранения в бд и на диск. То, что написано в этом методе, должно быть в одной из реализаций dao, в другой реализации dao будет сохранение в бд, но логика сервиса для них не отличается. Сервис подготавливает данные для сохранения (например, создает объект из 3 полей, проверяет, что файл корректный и все такое) и просто отдает данные дао.

Pelevina replied:
Сервисы сохраняют только в бд. В обоих сервисах одинаковая логика. Переделать один из них на сохранение на диск?
Использовать один класс-наследник от AbstractSequence, чтобы избавится от дубликатов кода?

darina-ovis replied:
https://itanddigital.ru/webapplications посмотрите раздел "Уровни архитектуры веб-приложений". Работа с бд и файловой системой - это уровень доступа к данным. На уровне сервиса у нас есть некий интерфейс, который может дать нам объект с файлом, accession и version. Откуда объект берется - из файловой системы или из бд, сервис не интересует. 

Насчет дубликтов. Вы можете использовать паттерн Template, например.

Pelevina replied:
Дубликаты удалены. Сохранение в бд сделать в отдельном методе или оставить в методе upload? Сервисы сделаны так, что работают с файлом из любого места, и дают accession и version

Pelevina committed:
Update FileUploadController.java

lyzhib committed:
Update DNAFileUploadService.java

lyzhib committed:
Update ProteinFileUploadService.java

lyzhib committed:
Update DNAFileUploadService.java

lyzhib committed:
Update DNAFileUploadService.java

lyzhib committed:
Update ProteinFileUploadService.java

iporsh commented on src/main/java/com/plasmidEditor/sputnik/controllers/FileUploadController.java:
```diff
@@ -0,0 +1,53 @@
+		proteinFileUploadService.upload(file);
+	}
+	
+	@ExceptionHandler(Exception.class)
```
Лучше вынести отдельно это , а то при появлении новых котроллеров, это часть не будет работать

Pelevina replied:
То есть в новый класс?

darina-ovis replied:
Если вы объявляете ExceptionHandler в контроллере, он действует только для этого класса. Если выносите в класс с аннотацией ControllerAdvice, то такой exception handler работает для всех контроллеров, где нет своих exception handler-ов

Pelevina replied:
А если нужно своё сообщение для каждого контроллера, что тогда делать?

darina-ovis replied:
Если в controller-е есть свой exception handler, он используется вместо controller advice

Pelevina replied:
Создан класс с аннотацией ControllerAdvice

iporsh commented on src/main/java/com/plasmidEditor/sputnik/uploadServices/DNAFileUploadService.java:
```diff
@@ -0,0 +1,55 @@
+
+public class DNAFileUploadService implements FileUploadService<DNASequence> {
+
+	@Override
```
код стайл(

Pelevina replied:
А что за замечания по код стайлу?

darina-ovis replied:
отступ съехал

Pelevina replied:
Выровнялось

iporsh commented on src/main/java/com/plasmidEditor/sputnik/uploadServices/DNAFileUploadService.java:
```diff
@@ -0,0 +1,55 @@
+//             if (genBankService.getByAccessionVersion(accession, version) == null) {
+//                 genBankService.upload(accession, version, new String(multipartFile.getBytes()));
+//             }
+            System.out.println("Upload " + accession + "." + version);
```
в консоль кидать ничего не стоит
стоит использовать лог

Pelevina replied:
В консоль в этом файле уже больше не кидает. Сохраняет в базу данных

iporsh commented on src/main/java/com/plasmidEditor/sputnik/uploadServices/DNAFileUploadService.java:
```diff
@@ -0,0 +1,55 @@
+            case 1:
+                break;
+            default:
+                throw new Exception("Upload " +
```
Лучше иметь кастомную ошибку

Pelevina replied:
Что такое кастомная ошибка?

darina-ovis replied:
значит, Exception, который вы написали сами, а не один из java пакетов

Pelevina replied:
Валидация находится на стороне ридера и использует пользовательское исключение

darina-ovis replied:
Не очень поняла ваш коммент. Exception - это слишком общее понятие. Всегда лучше быть более конкретным. Даже если вы не хотите создавать свой Exception, лучше бросать что-то более менее специфичное, например IllegalArgumentException
https://programming.guide/java/list-of-java-exceptions.html вот полный список - посмотрите что лучше по смыслу вам подходит. Или создайте что-то свое

Pelevina replied:
Я имела ввиду, что теперь используется более специфичное исключение, чем Exception

iporsh commented on src/main/java/com/plasmidEditor/sputnik/uploadServices/ProteinFileUploadService.java:
```diff
@@ -0,0 +1,51 @@
+        }
+    }
+
+    private void validateListSize(List<ProteinSequence> sequences) throws Exception {
```
можем мы вынести это в отдельный класс т.к. код задублирован (точно также есть как минимум в DNAFileUploadService )

Pelevina replied:
Проверка размеров находится на стороне ридера

Pelevina committed:
Update DNAFileUploadService.java

Pelevina committed:
Create FileUploadException.java

Pelevina committed:
Update DNAFileUploadService.java

lyzhib committed:
Update ProteinFileUploadService.java

darina-ovis commented on src/main/java/com/plasmidEditor/sputnik/uploadServices/DNAFileUploadService.java:
```diff
@@ -0,0 +1,46 @@
+				genBankServiceImpl.save(new GenBankDTO(accession, String.valueOf(version), content));
+			}
+			else {
+				throw new NullPointerException();
```
Не думаю, что это хорошая идея кинуть тут NPE. Вы ее поймаете в 40 строчке и выведете пустое сообщение.

Pelevina replied:
Чем тогда лучше заменить эту инструкцию в 38 строке? Ведь нужно, чтобы если sequence является nullом, то выходило исключение.

darina-ovis replied:
Если что-то не должно быть null, нужно проверить, что оно не null, а если null, то выдать осмысленное сообщение. В вашем коде объект даже не null, он экземпляр другого класса. NPE как-то совсем не про это. 

Вот например в библиотеке guava есть такие методы 

https://guava.dev/releases/20.0/api/docs/src-html/com/google/common/base/Preconditions.html#line.120

```
public static void checkArgument(boolean expression, @Nullable Object errorMessage) {
    if (!expression) {
      throw new IllegalArgumentException(String.valueOf(errorMessage));
    }
  }
```

вызывается он так `Preconditions.checkArgument(sequence instanceof DNASequence, "DNASequence is expected")`

Вам не нужно использовать тут guava. Это просто пример

Pelevina replied:
Изменён тип исключения. Добавлено сообщение

darina-ovis commented on src/main/java/com/plasmidEditor/sputnik/uploadServices/ProteinFileUploadService.java:
```diff
@@ -0,0 +1,46 @@
+		ProteinSequence sequence;
+
+        	try {
+            		File file = File.createTempFile("Protein", ".tmp");
+            		file.deleteOnExit();
+            		inputStream.transferTo(file);
+            		sequence = new FileProteinGenbankManager().readSequense(file.getAbsolutePath());
+			if (sequense instanceof ProteinSequence) {
+        			String accession = sequence.getAccession().getID();
+        			int version = sequence.getAccession().getVersion();
+				String content = new ReaderUtils().readStringFromFile(file.getAbsolutePath());
+				genBankServiceImpl.save(new GenBankDTO(accession, String.valueOf(version), content));
+			}
+			else {
+				throw new NullPointerException();
+			}
+        	} catch (Exception e) {
+            		throw new FileUploadException(e.getMessage(), e);
+        	}
```
постарайтесь избавиться от дубликатов кода. Сейчас этот код очень похож на DNAFileUploadService. И пишется sequen**C**e - грамматика очень важна для кода. Потому что когда проект большой, программисты часто ищут что-то по какому-нибудь слову. Если слово написано с ошибкой, то это место могут не найти и это может быть причиной ошибки в будущем. Обычно идея подсвечивает такие слова

Pelevina replied:
Sequence исправлено.

darina-ovis replied:
Исправьте во всех местах) Когда код пишут методом copy/paste, очень часто забывают про все места - тем и страшны дубликаты кода

Pelevina replied:
Дубликаты удалены

Pelevina committed:
Update DNAFileUploadService.java

Pelevina committed:
Update ProteinFileUploadService.java

Pelevina committed:
Update FileUploadController.java

Pelevina committed:
Туда вынесен дублирующийся код

Pelevina committed:
Избавление от дублирующегося кода

Pelevina committed:
Избавление от дублирующегося кода

Pelevina committed:
Update FileUploadService.java

Pelevina committed:
Update DNAFileUploadService.java

Pelevina committed:
Update FileUploadServiceUtils.java

Pelevina committed:
Update FileUploadException.java

Pelevina committed:
Добавление общей обработки

Pelevina committed:
Update FileUploadController.java

iporsh commented on src/main/java/com/plasmidEditor/sputnik/exceptions/FileUploadException.java:
```diff
@@ -0,0 +1,7 @@
+package com.plasmidEditor.sputnik.exceptions;
+
+public class FileUploadException extends Exception {
```
А почему Exception ?

Pelevina replied:
Exception - базовый класс для всех видов exception. Какой класс лучше всего подойдёт в качестве родителя FileUploadException?

Pelevina replied:
Переделано на RuntimeException

iporsh commented on src/main/java/com/plasmidEditor/sputnik/controllers/JsonError.java:
```diff
@@ -0,0 +1,10 @@
+
+@Data
+@AllArgsConstructor
+public class JsonError {
```
странное название

Pelevina replied:
В чём его странность? Как лучше переименовать этот класс?

Pelevina replied:
Переименовано в ErrorResponseBody

iporsh commented on src/main/java/com/plasmidEditor/sputnik/uploadServices/FileUploadServiceUtils.java:
```diff
@@ -0,0 +1,38 @@
+
+public class FileUploadServiceUtils<T extends AbstractSequence> {
+	public File writeToTmpFile(InputStream inputStream) throws IOException {
+		File file = File.createTempFile("dna", ".tmp");
```
почему именно эти рассширения?
Можем мы их вынести в настройку?

Pelevina replied:
Расширение AbstractSequence, потому что тип T должен быть наследником класса AbstractSequence.

Pelevina replied:
Создания временного файла больше нет.

Pelevina committed:
Update FileUploadException.java

Pelevina committed:
Update and rename JsonError.java to ErrorResponseBody.java

Pelevina committed:
Update ControllersExceptionHandler.java

Pelevina committed:
Update FileUploadController.java

Pelevina committed:
Update FileUploadServiceUtils.java

darina-ovis commented on src/main/java/com/plasmidEditor/sputnik/uploadServices/ProteinFileUploadService.java:
```diff
@@ -0,0 +1,46 @@
+
+public class ProteinFileUploadService implements FileUploadService<ProteinSequence> {
+
+	private final GenBankServiceImpl genBankServiceImpl;
```
это должен быть интерфейс уровня дао

Pelevina replied:
А это есть реализация интерфейса уровня дао

darina-ovis replied:
```suggestion
	private final GenBankService genBankService;
```

Pelevina replied:
Изменила с GenBankServiceImpl на GenBankService

Pelevina committed:
Добавление констант для FileUploadController

Pelevina committed:
Константы вместо строк

Pelevina committed:
Update FileUploadServiceUtils.java

Удаление создания временного файла, недопуск повторной загрузки файла, добавка валидации последовательности.

Pelevina committed:
Update DNAFileUploadService.java

Pelevina committed:
Update ProteinFileUploadService.java

Pelevina committed:
Update DNAFileUploadService.java

Pelevina committed:
Update FileUploadServiceUtils.java

Pelevina committed:
Add files via upload

darina-ovis commented on src/main/java/com/plasmidEditor/sputnik/uploadServices/FileUploadServiceUtils.java:
```diff
@@ -0,0 +1,46 @@
+		}
+	}
+	
+	public void saveSequenceToDB(T sequence, InputStream inputStream, GenBankServiceImpl genBankServiceImpl) throws IllegalArgumentException, 
```
Это выглядит странно. У вас всего один GenBankServiceImpl - зачем его передавать параметром? Классы типа Utils обычно содержат статические методы. Многие программисты складывают туда все, что не знают куда положить. Это плохая практика. Разберитесь со структурой. Если вам приходится передавать в метод ссылку на объект только потому, что этот объект бин, значит, что-то пошло не так.

Pelevina replied:
Те методы класса GenBankServiceImpl, что я применяла, не статические и написаны другими ребятами. Нужно добавить модификатор static в эти методы?

Pelevina replied:
Теперь есть абстрактный класс вместо интерфейса и утильного класса

darina-ovis commented on src/main/java/com/plasmidEditor/sputnik/controllers/FileUploadController.java:
```diff
@@ -0,0 +1,48 @@
+	
+	@ExceptionHandler(Exception.class)
+	public ResponseEntity<JsonError> handleException(Exception e) {
+		return ResponseEntity.badRequest().body(new ErrorResponseBody("failed to upload genbank file", e.getMessage()));
```
при таком подходе в лог никогда не попадет stacktrace и будут чистые логи, и только пользователи будут видеть ошибку.
Можете не исправлять в данном случае, просто помните, что выводить в логи ошибки очень важно в боевых проектах.

Pelevina committed:
Update DNAFileUploadService.java

Pelevina committed:
Update FileUploadServiceUtils.java

Pelevina committed:
Update ProteinFileUploadService.java

Pelevina committed:
Update ControllersExceptionHandler.java

Pelevina committed:
Update FileUploadController.java

Commit:
remove FileUploadServiceUtils class

darina-ovis commented on src/main/java/com/plasmidEditor/sputnik/uploadServices/FileUploadService.java:
```diff
@@ -0,0 +1,36 @@
+    public abstract void upload(InputStream inputStream) throws FileUploadException;
+
+    protected void saveSequenceToDB(T sequence, InputStream inputStream) throws IllegalArgumentException, IOException, GenBankNotFoundException, RepeatFileUploadingException {
+        GenBankServiceImpl genBankServiceImpl = new GenBankServiceImpl();
```
Это Spring bean  - его не надо создавать, его надо inject-тить, т.е. должно быть так
```
        @Autowired
        GenBankService genBankService;
        
        protected void void saveSequenceToDB(T sequence, InputStream inputStream) throws IllegalArgumentException, IOException, GenBankNotFoundException, RepeatFileUploadingException {
        ...
        if(genBankService.is...)
   
```

Обратите внимание, что лучше объявлять его как interface, не конкретный класс. Spring сам разберется и найдет подходящую реализацию интерфейса (в вашем случае она одна)

Pelevina replied:
Добавлено Autowired поле типа GenBankService и используется интерфейс.

Pelevina committed:
Update FileUploadController.java

Pelevina committed:
Добавка Autowired поля типа GenBankService

Commit:
Merge branch 'sputnik' into sputnik_upload

# Conflicts:
#	src/main/java/com/plasmidEditor/sputnik/services/GenBankService.java

Commit:
Merge remote-tracking branch 'origin/sputnik_upload' into sputnik_upload
